allow focus for user list

hzulla opened the pull request:
In `user_list.py`, this replaces the EventBox widget with a KanoButton widget.

By doing this, `kano-greeter` gains keyboard focus for the user list, making the login process easier to use without a mouse.

**BUT: There is still one bug** I couldn't resolve today and that's why this PR is still incomplete.

The font setting for the CSS rule `.user GtkLabel` is being ignored and Gtk shows the default CSS font setting for KanoButton instead (Bariol 13, I guess?).

The CSS rule is okay: adding `background: green;` to the rule will paint the GtkLabel green, so it *is* the right CSS path.

Fiddled around with the CSS but no success, Gtk CSS is strange. Any idea about how to define the font size in the CSS file so that Gtk actually applies it? Thanks.

hzulla commented:
Btw, later I will add a `grab_focus()` to the user who was last logged in on the system.

hzulla commented:
Update - now implemented the actual reason for this PR: `kano-greeter` will remember the username for the last login on this machine and then set the initial focus to that user in `user_list.py`.

As a result, it's now possible to login without using the mouse.

tombettany commented:
@hzulla Interesting work. Good idea to use the Kano button. I've not had a chance to test this yet but I do have a few comments on the code. I'll let you know as soon as I've had an opportunity to actually run it.

tombettany commented on kano_greeter/last_user.py:
```diff
@@ -0,0 +1,22 @@
+    try:
+        with open(os.path.join(LAST_USER_PATH, LAST_USER_FILE)) as f:
+            last_user = f.read()
+            f.close()
```
We are using Python 2.7, we shouldn't need to close the file.

tombettany commented on kano_greeter/last_user.py:
```diff
@@ -0,0 +1,22 @@
+def set_last_user(last_user):
+    if not os.path.exists(LAST_USER_PATH):
+        os.makedirs(LAST_USER_PATH)
+    f = open(os.path.join(LAST_USER_PATH, LAST_USER_FILE), "w+")
```
Can this become a `with open(...) as f:` instead?

tombettany commented on kano_greeter/last_user.py:
```diff
@@ -0,0 +1,22 @@
+import os
+
+LAST_USER_PATH = '/var/lib/lightdm/.cache/kano-greeter/'
+LAST_USER_FILE = 'last-user'
```
Would it make sense to change `LAST_USER_FILE` to be `os.path.join(LAST_USER_PATH, 'last-user')` instead to simplify the file IO calls later?

hzulla replied:
True, will change that.

I wasn't sure about the path to use. /var/lib/lightdm/.cache/ is what lightdm-gtk-greeter uses, but since kano-greeter comes as a separate project, maybe /var/lib/kano-greeter is more appropriate?

tombettany replied:
Hmmm, I'm not really sure what the path should be. @pazdera any preference?

tombettany commented on kano_greeter/user_list.py:
```diff
@@ -70,32 +75,19 @@ def add_account(*_):
 
     def __init__(self, username):
-        Gtk.EventBox.__init__(self)
+        KanoButton.__init__(self, text=username.title(), color="orange")
```
We may not be very good at it but let's try to be consistent with `'` and `"`. Nearly all of the current code uses `'` so we'll keep that as the preference here.

At some point we should also change the couple of other cases where `"` is used.

hzulla replied:
True. Are there any sane pythonesque rules for when to choose which or do you guys have an internal rule? I'm from the Perl world, where the quotes have different meanings, and [people seem to agree](https://www.python.org/dev/peps/pep-0008/#string-quotes) that the following is a good rule (at least it makes sense to me, too):

> I like to use double quotes around strings that are used for interpolation or that are natural language messages, and single quotes for small symbol-like strings.

On the other hand, the Python guys themselves [don't recommend one over the other](https://www.python.org/dev/peps/pep-0008/#string-quotes).

tombettany commented on kano_greeter/password_view.py:
```diff
@@ -86,6 +87,8 @@ def _authentication_complete_cb(self, _greeter):
             .format(self.user))
 
+        set_last_user(self.user)
+        
```
Blank line should have no whitespace on it

tombettany commented on kano_greeter/user_list.py:
```diff
@@ -35,6 +36,8 @@ def __init__(self):
 
         title = Heading('Select Account', 'Log in to which account?')
         self.box.pack_start(title.container, False, False, 0)
+        
```
Trailing whitespace again

hzulla commented:
Thx. Will fix.

hzulla commented:
ok, fixed.

Orange KanoButton was only used because it comes with working support for keyboard focus, incl. style changes (dark border for focussed item) when needed.

* code style fixed as per request
* `last-user` path has changed (but if you use this patch, you still might decide to put it elsewhere)
* added user name above password entry widget (meant as visual feedback to user when accidentally trying to login for the wrong username)
* not implemented: it might be beneficial to catch the ESC-key on the password view, so that one can return to the user list without the top_bar arrow, which requires clicking with a mouse
* not fixed: Gtk CSS still doesn't use the `.user GtkLabel` font rule, it's overridden by the KanoButton font defaults. No idea how to fix that.

tombettany commented on kano_greeter/last_user.py:
```diff
@@ -0,0 +1,21 @@
+
+import os
+
+LAST_USER_PATH = '/var/lib/kano-greeter'
```
@hzulla I've just tried this and I get a permission error when attempting to write to `/var/lib/kano-greeter`. The greeter is not run as `root` but by a restricted `lightdm` user whose home directory is `/var/lib/lightdm`. The easiest solution might be to just change `LAST_USER_PATH` to be `/var/lib/lightdm/kano-greeter` or `/var/lib/lightdm/.cache/kano-greeter` (as you suggested earlier).

hzulla replied:
Darn. I only tried it locally. Sorry.

tombettany commented:
I manually changed the `LAST_USER_PATH` to a subdirectory of `/var/lib/lightdm` and everything worked a charm. Once this has been changed it is good to merge.

hzulla commented:
Use `/var/lib/lightdm/.cache/kano-greeter` instead, please. That seems to be what other greeters are using.

tombettany commented:
@hzulla That is fine by me. Are you okay to commit the change or would you like me to do it?

hzulla commented:
@tombettany path changed by last commit.

But do you have any idea how to fix the CSS issue? I'm not able to change the font for the orange KanoButton to Bariol Bold 18, the CSS rule is being ignored. (See earlier comment.)

tombettany commented:
@hzulla Sorry, I missed your comment there.

Having spoken to @carolineclark it looks like it a problem that we've been having with the implementation of all the standard buttons. 

Gtk3 uses a set of `Style Provider Priorities` which can be used when applying the styles (https://github.com/KanoComputing/kano-toolset/blob/master/kano/gtk3/apply_styles.py#L44-L53). Unfortunately it looks like the only one that we could get to display the Kano Button consistently was the highest priority - `STYLE_PROVIDER_PRIORITY_USER` (we [apply styles](https://github.com/KanoComputing/kano-toolset/blob/master/kano/gtk3/buttons.py#L78) which [adds the highest priority](https://github.com/KanoComputing/kano-toolset/blob/master/kano/gtk3/apply_styles.py#L67)) - which stops us being able to override it.

It isn't really necessary to have the font size 18 for the greeter buttons because we actually want the consistency brought by using the `KanoButton`. I propose that we just remove [the font CSS](https://github.com/hzulla/kano-greeter/blob/c02aafcc84a74cedf636750b79d0454a141b1617/kano_greeter/media/css/kano-greeter.css#L4-L7) and use the default for consistency.

tombettany commented:
Perfect, now merging.

hzulla commented:
thx.
